### PR TITLE
Make AltServer menu appear attached to the icon

### PR DESCRIPTION
**Pull request by tbodt**
Also a bit of a general cleanup.

### CODE REVIEW

File: AltServer/AppDelegate.swift [hunk 1]
```diff
@@ -68,28 +62,6 @@ class AppDelegate: NSObject, NSApplicationDelegate {
...
-    {
-        guard let button = self.statusItem?.button, let superview = button.superview, let window = button.window else { return }
-        
-        self.connectedDevices = ALTDeviceManager.shared.connectedDevices
-        
-        self.launchAtLoginMenuItem.state = LaunchAtLogin.isEnabled ? .on : .off
-        self.launchAtLoginMenuItem.action = #selector(AppDelegate.toggleLaunchAtLogin(_:))
-        
-        let x = button.frame.origin.x
-        let y = button.frame.origin.y - 5
```
Comment (chickdan) [inline AltServer/AppDelegate.swift]: I think you could have simply removed the `-5` here to make it appear attached without making the rest of these changes.

File: AltServer/AppDelegate.swift [hunk 2]
```diff
@@ -37,16 +37,10 @@ class AppDelegate: NSObject, NSApplicationDelegate {
...
         
         let item = NSStatusBar.system.statusItem(withLength: -1)
-        guard let button = item.button else { return }
-        
-        button.image = NSImage(named: "MenuBarIcon")
-        button.target = self
-        button.action = #selector(AppDelegate.presentMenu)
-        
+        item.menu = self.appMenu
+        item.button!.image = NSImage(named: "MenuBarIcon")
```
[the remark this answers is not in the document]
Comment (tbodt) [reply AltServer/AppDelegate.swift]: I have no idea when this would ever possibly be nil, but I can revert this if you want.

File: AltServer/AppDelegate.swift (hunk 1 above)
Comment (tbodt) [reply AltServer/AppDelegate.swift]: No, that doesn't do the same thing. I'll upload screenshots in a bit.

**Comment by tbodt** [comment]
> In general, I disagree with these changes because it removes clearly defined behavior and masks it behind "automatic" behavior generated by the storyboard (I'm assuming that's what's happening here because this PR is mostly deletion and very little addition). Yes, it's a little more boilerplate-y, but it allows developers to easily jump in and see the intention of others. In general I prefer programmatic UI over storyboard because it's so much easier for collaborative development.

That's actually not what's happening here. The code deleted from the storyboard was all dead code, it defined the app menu bar which obviously doesn't exist for AltStore. The code deleted in AppDelegate is all replaced with that one `item.menu = self.appMenu` line, which makes the menu automatically pop up when you click the status item.

File: AltServer/AppDelegate.swift [hunk 3]
```diff
@@ -198,11 +170,21 @@ extension AppDelegate: NSMenuDelegate
 {
     func numberOfItems(in menu: NSMenu) -> Int
     {
+        if menu != self.connectedDevicesMenu
+        {
+            return -1
```
[the remark this answers is not in the document]
Comment (tbodt) [reply AltServer/AppDelegate.swift]: No. The documentation for numberOfItems says
> If you return a negative value, the number of items is left unchanged and menu(_:update:at:shouldCancel:) is not called.

File: AltServer/AppDelegate.swift (hunk 1 above)
Comment (tbodt) [reply AltServer/AppDelegate.swift]: master:

Screen Shot 2020-01-07 at 8 59 40 PM

master with the -5 removed:

Screen Shot 2020-01-07 at 9 13 27 PM

This PR:

Screen Shot 2020-01-07 at 9 16 27 PM

First two definitely look wrong to me.

**Comment by chickdan** [comment]
Removing the dead code from the storyboard file is fine and not what my comment was related to.

> The code deleted in AppDelegate is all replaced with that one item.menu = self.appMenu line, which makes the menu automatically pop up when you click the status item.

What you are saying isn't happening is what is actually happening. The removal of the code from the AppDelegate _is_ moving codified behavior to the "automatic" behavior of the storyboard. `appMenu` is an `IBOutlet`, ie, a storyboard connection. I'm not saying it's wrong, just giving my opinion because I've had miserable experiences working with storyboards in a multi-developer environment.

**Comment by tbodt** [comment]
The app menu was in the storyboard before this, it's just the code that displays it now is different. Are you asking me to move it out into code?

File: AltServer/AppDelegate.swift [hunk 4]
```diff
@@ -222,6 +195,15 @@ extension AppDelegate: NSMenuDelegate
         
         return true
     }
+    
+    func menuWillOpen(_ menu: NSMenu)
+    {
+        if menu == self.appMenu
+        {
+            self.connectedDevices = ALTDeviceManager.shared.connectedDevices
+            self.launchAtLoginMenuItem.state = LaunchAtLogin.isEnabled ? .on : .off
```
Comment (ivar1803) [inline AltServer/AppDelegate.swift]: We can remove self prefixes here. 

if menu == appMenu {
    connectedDevices = ALTDeviceManager.shared.connectedDevices

**Comment by C0braD3v** [comment]
This PR is pretty good... makes the button not overlap

**Comment by tbodt** [comment]
@chickdan Please review, there's no longer any code added to the IB file

**Comment by chickdan** [comment]
Thanks for keeping it codified! It's appreciated!

**Comment by C0braD3v** [comment]
Overall I feel like these changes don’t offer any major benefits...

**Comment by tbodt** [comment]
It's not a major thing, just a detail that bugs me.

**Comment by C0braD3v** [comment]
Any reason this PR is aimed at the master branch and not the Develop branch? Pretty sure all PRs and changes would go on develop.

**Comment by tbodt** [comment]
That would be because I didn't know PRs should go against develop.

**Comment by tbodt** [comment]
Now made against develop

File: AltServer/AppDelegate.swift [hunk 5]
```diff
@@ -67,14 +67,11 @@ class AppDelegate: NSObject, NSApplicationDelegate {
...
         
         let item = NSStatusBar.system.statusItem(withLength: -1)
-        guard let button = item.button else { return }
-        
-        button.image = NSImage(named: "MenuBarIcon")
-        button.target = self
-        button.action = #selector(AppDelegate.presentMenu)
-        
+        item.menu = self.appMenu
+        item.button!.image = NSImage(named: "MenuBarIcon")
```
Comment (HPaulson) [inline AltServer/AppDelegate.swift:71]: Unsure about this.. I don't this this would ever be nil, No need for revert Imo

**Comment by tbodt** [comment]
@rileytestut

**Comment by MaivKhanh20181** [comment]
1

**Comment by MaivKhanh20181** [comment]
1

File: AltServer/AppDelegate.swift [hunk 6]
```diff
@@ -479,13 +433,47 @@ private extension AppDelegate
...
+        
         return self.connectedDevices.isEmpty ? 1 : self.connectedDevices.count
     }
     
     func menu(_ menu: NSMenu, update item: NSMenuItem, at index: Int, shouldCancel: Bool) -> Bool
     {
+        if menu != self.connectedDevicesMenu
+        {
+            return false
+        }
```
Comment (rileytestut) [inline AltServer/AppDelegate.swift:475]: Another `if` -> `guard` 

```suggestion
        guard menu == self.connectedDevicesMenu else { return }
```

File: AltServer/AppDelegate.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (rileytestut) [reply AltServer/AppDelegate.swift:71]: Since I'm already using `guard-let`, I don't see a reason to change it to be force unwrap.

File: AltServer/AppDelegate.swift [hunk 7]
```diff
@@ -479,13 +433,47 @@ private extension AppDelegate
 
 extension AppDelegate: NSMenuDelegate
 {
+    
+    func menuWillOpen(_ menu: NSMenu)
+    {
+        if menu == self.appMenu
```
Comment (rileytestut) [inline AltServer/AppDelegate.swift:439]: Can you turn this `if` into a `guard`?

```
guard menu == self.appMenu else { return }
```

File: AltServer/AppDelegate.swift [hunk 8]
```diff
@@ -479,13 +433,47 @@ private extension AppDelegate
...
+            self.installMailPluginMenuItem.action = #selector(AppDelegate.handleInstallMailPluginMenuItem(_:))
+        }
+    }
+
     func numberOfItems(in menu: NSMenu) -> Int
     {
+        if menu != self.connectedDevicesMenu
+        {
+            return -1
+        }
```
Comment (rileytestut) [inline AltServer/AppDelegate.swift:465]: Same here, generally I prefer `guard` for early returns unless the logic is hard to follow.

```suggestion
        guard menu == self.connectedDevicesMenu else { return -1 }
```

File: AltServer/AppDelegate.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (tbodt) [reply AltServer/AppDelegate.swift]: Returning if item.button is nil doesn't seem right to me, since it silently skips the rest of the method which seems to be doing important things. What do you think of:
```suggestion
	if let button = item.button
	{
	        button.image = NSImage(named: "MenuBarIcon")
	}
```

File: AltServer/AppDelegate.swift [hunk 9]
```diff
@@ -67,14 +67,15 @@ class AppDelegate: NSObject, NSApplicationDelegate {
...
-        button.image = NSImage(named: "MenuBarIcon")
-        button.target = self
-        button.action = #selector(AppDelegate.presentMenu)
-        
+        item.menu = self.appMenu
+        item.button?.image = NSImage(named: "MenuBarIcon")
+        if let button = item.button
+        {
+            button.image = NSImage(named: "MenuBarIcon")
+        }
```
Comment (rileytestut) [inline AltServer/AppDelegate.swift:75]: Redundant image assignment? I agree early return doesn't make as much sense here, so I'd prefer just the optional chaining.
```suggestion
        item.button?.image = NSImage(named: "MenuBarIcon") 
```

Comment (tbodt) [reply AltServer/AppDelegate.swift:75]: Oops, definitely didn't mean to assign the image twice

**Comment by rileytestut** [comment]
Thanks for your changes! Everything looks good 👍
